CIP-????/README.md

```diff
@@ -0,0 +1,530 @@
+
+#### Block
+
+The cddl for a block is currently:
+
+```cddl
+block =
+  [ header
+  , transaction_bodies         : [* transaction_body]
+  , transaction_witness_sets   : [* transaction_witness_set]
+  , auxiliary_data_set         : {* transaction_index => auxiliary_data }
+  , invalid_transactions       : [* transaction_index ]
+  ]; Valid blocks must also satisfy the following two constraints:
+   ; 1) the length of transaction_bodies and transaction_witness_sets
+   ;    must be the same
+   ; 2) every transaction_index must be strictly smaller than the
+   ;    length of transaction_bodies
+```
+
+Just like with the transaction cddl, the block's transaction bodies and the transaction witness sets
+will be merged, and the constraints must be updated accordingly:
+
+```cddl
+; This is re-posted here for reference.
+transaction_piece =
+  { 0: transaction_body
+  , 1: transaction_witness_set ; The witness set for this transaction body.
+  }
+
+; A specific transaction in the block is a specific batch of transaction pieces. 
+block_transaction = [* transaction_piece]
+
+block =
+  [ header
+  , block_transactions   : [* block_transaction] ; This is new
+  , auxiliary_data_set   : {* transaction_index => auxiliary_data }
+  , invalid_transactions : [* transaction_index ]
+  ]; Valid blocks must also satisfy the following constraint:
+   ; 1) every transaction_index must be strictly smaller than the
+   ;    length of block_transactions
+```
+
+### Transaction Size Limit
+
+A transaction broken into pieces should still have approximately the same total size as the whole
+transaction. There are a few things to note:
+
+- If two transaction pieces require the same witness (ie, both pieces require a signature from
+Alice), this witness would now effectively appear twice.
+- Certain items can appear in multiple pieces (eg, the same input can be spent/referenced, tokens can
+be minted/burned that would otherwise cancel out, etc).
+
+Despite the above, it is probably fine to leave the maximum transaction size as is, at least to
+start. The transaction fee would naturally incentivize users/aggregators to minimize the amount of
+duplication in a transaction.
+
+### Block Size Limit
+
+Due to the possible duplicate items in each transaction and the fact that a block is made up of
+multiple transactions, the issues in the Transaction Size Limit section are slightly exacerbated.
+However, it should still be fine to keep the current block size limit, at least initially. Just like
+with transactions, the transaction fee would naturally incentivize users to minimize duplications.
+
+### Smart Contract Execution
+
+While the cryptographic keys only need to sign the relevant transaction pieces, the smart contracts
+would be executed against the full transaction. This requirement enables the current smart contract
+interface to remain mostly untouched. The flow is essentially:
+
+1. Deterministically derive the whole transaction from the list of pieces.
```

Comment:
> - two transactions that withdraw from the same reward account. It is irrelevant that one of them will be zero, it will change semantics. Moreover if a stake credential is a scritp, then how can you ignore running one of them.

I would argue this transaction should be rejected. I think you are over-complicating things. IMO things should either exact match or be rejected. The only exceptions are when there is a very obvious way to merge them, like with mints and validity intervals. 

> - A DRep casting a yes vote on a proposal in one transaction and then a No on the same proposal in a separate transaction. You might think that we could then ignore the first vote, since the second one replaces the first. But, what if the DRep is backed by a script and it would actually fail if we didn't ignore it.

This transaction would be rejected as well because the proposal vote in one piece does not exact match the proposal vote in the other. The pieces are not applied to the ledger sequentially. They are folded into a single transaction and applied as a unit.

> - Two transactions minting the same asset. How do you run the script twice? You'll be forced to put squash those two mints together into one, which could affect the execution of the script.

The mints would be merged upon folding the pieces and the script would only be executed once. As I said before, smart contracts are a weak-point with this CIP. AFAIU, the new Validation Zone revision addresses this limitation by actually having smart contracts validate against the piece they are used in.

Reply:
I am not over-complicating! Those are all of the cases that would have to be addressed in the ledger rules. If we are to reject them, we need to implement specific predicate failures for each one of them that do the rejection. I can see this CIP adding more predicate failures than the total number of predicate failures that we already have today.

> Validation Zone revision addresses this limitation 

I am really confused here. [Validation Zones](https://github.com/cardano-foundation/CIPs/pull/862) is a totally different CIP. Are you trying to combine the two together somehow?

The only way I currently see combining multiple transactions into one for validation to work is if we restrict all those pieces to be **completely** disjoined, i.e. no common credentials.

Reply:
> I am not over-complicating!

Sorry, my confusion came from not realizing the redeemers included the execution budgets and therefore you can't do an exact match on the script_data_hash across pieces. I was under-complicating it :sweat_smile:

> I am really confused here. Validation Zones is a totally different CIP. Are you trying to combine the two together somehow?

The new Validation Zone revision (the [README-implicit.md](https://github.com/cardano-foundation/CIPs/blob/1fd1d6dc56fadca8b07009e795b3ec4eedc298cd/CIP-0118/README-implicit.md)) is trying to incorporate the unbalanced transaction idea from this CIP. It uses them in an entirely different way than how this CIP used them.